### ISSUE TITLE
[Bug] Multiple MM2 worker nodes cause missing topics and configuration updates

### ISSUE THREAD

**Issue by AljoschaP** [strimzi/strimzi-kafka-operator#3688]
**Describe the bug**
If you increase the Mirrormaker2 replicas to more than one node, the task config can't be updated by the connector (https://issues.apache.org/jira/browse/KAFKA-9981). This can cause missing topics because the tasks.assigned.partitions also got no updates. 

**To Reproduce**
1. Start MM2 with 3 replicas
2. Create new topic after MM2 has been started
3. Topic will not show up in tasks.assigned.partitions
4. Change connector configuration
5. Configuration changed for connector but not for task

**Expected behavior**
The tasks.assigned.partitions should be updated when a new topic got created and the config of the task should be updated when the config of the connector got updates.

**Environment (please complete the following information):**
 - Strimzi version: 0.18.0
 - Installation method: YAML Files
 - Kubernetes cluster: Kubernetes 1.17.9
 - Infrastructure: AKS

**Additional context**
I know that this is not a bug of Strimzi but it can cause vicious headaches for new users and huge problems in production if not recognized. Maybe the Kafka bug can be mitigated by restricting the replicas to one until the original bug is fixed.

**Comment by scholzj** [comment]
Are you sure this bug is related to Strimzi's implementation of MM2? We do not use the `connect-mirror-maker.sh`. We basically deploy Connect and on top of it the MirrorMaker2 connectors. So it has the things such as the REST API etc. I know nothing about the MM2 internals - but the issue description makes me think Strimzi should not be impacted by this.

@ajborley Any thoughts?

**Comment by ajborley** [comment]
@scholzj yep, agreed - https://issues.apache.org/jira/browse/KAFKA-9981 appears to only affect running MM2 in distributed cluster mode, not via a Kafka Connect instance.

Re. your reproduction steps, what is your `refresh.topics.internval.seconds` config for the MirrorSourceConnector set to in your CR? If it's not set it will be the default of 10 minutes, so it can take a while for MM2 to notice new topics and partitions.

**Comment by AljoschaP** [comment]
@ajborley It is the default value (10 Minutes) and the topic was created in the backup cluster. So this seems to work. The problem was that the replication didn't happen because the source connector task didn't pick up the new topic as part of the tasks.assigned.partitions. Even after 24 hours.

**Comment by feydan** [comment]
I'm experiencing something maybe tangential.  When I set MirrorMaker2 replicas to more than 1, it doesn't seem to be honoring my tasksMax configuration.  In my instance, I set tasksMax to 6 for checkpointConnector with 3 mm2 replicas, and only two tasks were running (I let it run for 12+ hours with no change).  Once I set mm2 back to 1 replica, all 6 tasks came up immediately.

**Comment by OfirTadmor** [comment]
Hey, 
We experienced similar problems, 
either changing tasksMax or adding a new topic to the source cluster
is not updated (and topic not replicated to target) until we change replicas to 1.
Basically, it seems that config updates in a cluster MM2 are broken in some way.
The workaround we found is to manually set the replicas to 1 and then return it to the wanted number, as @feydan  suggested.

**Comment by AljoschaP** [comment]
I can confirm the observations of @OfirTadmor and @feydan for Kafka 2.5.0 and for Kafka 2.7.0 and I think that all of it is somehow related. It seems to me that there is something fundamentally broken with distributed MM2 in connect clusters and maybe dedicated clusters. The only thing I don't understand is that this error doesn't seem to occur in every setup. So there must be a perfect storm between Strimzi and MM2. Maybe its the way how the MM2 worker deployment is done. Because MM2 is a deployment and not an sts, all pods are created at once and maybe this creates some sort of deadlock. What I can say is that after multiple tries, some connectors randomly start working, creating tasks, start replicating and some just get stuck and do nothing until I downscale the replicas to 1. I'll try to verify my theory and give an update asap.

**Comment by AljoschaP** [comment]
Short update: I tried to simulate the OrderedReady pod management of STS by scaling the mm2 deployment as soon as the previous pod is ready and it seems that its much more stable. Before I try to start to working on it, I would like to know if its okay for you to change MM2 to a STS ? @ajborley @scholzj

**Comment by scholzj** [comment]
But it is not using StatefulSets. So you would need to clarify how you want to do it. It would also likely not involve only MM2 but also Connect since they share some code. So I think we need to first:
* Understand how you want to implement it
* Understand the actual reasons for it and possibly try to fixed them in Kafka instead
* If you think it is useful, it should probably also have a proposal (https://github.com/strimzi/proposals) to record how any why should it be done

**Comment by AljoschaP** [comment]
Okay :smiley:  Thanks for the advice :+1:

**Comment by scholzj** [comment]
@tombentley @mimaison Do you have any idea why this would make difference? ^^^

**Comment by AljoschaP** [comment]
I think, I found the error and @scholzj and @mimaison were even part of the review of the [fix](https://github.com/apache/kafka/pull/11174). After a lot more in-depth debugging it seems that the reconfiguration of tasks from a follower doesn't work because an unsafe url is passed to the rest client of the connect framework. 

```
           String serializedBody = requestBodyData == null ? null : JSON_SERDE.writeValueAsString(requestBodyData);
            log.trace("Sending {} with input {} to {}", method, serializedBody, url);

            Request req = client.newRequest(url);
            req.method(method);
            req.accept("application/json");
            req.agent("kafka-connect");
```
https://github.com/apache/kafka/blob/6eed7743ff6c0e73d65c09bac2e2ad9586cc56ce/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestClient.java#L107

And at this point the url is e.g.:

```
http://10.255.140.160:8083/connectors/kafka-source-cluster->kafka-backup-cluster.MirrorCheckpointConnector/tasks
```
which caused an `java.net.URISyntaxException` in the jetty http client. So no request is done and the reconfiguration fails. If this is right and I don't miss something, MM2 is not working with more than one worker until 2.8.2, right?

**Comment by scholzj** [comment]
Damn, I already forgot about that fix. Sorry. It seems like that was a long time ago :-/.

**Comment by AljoschaP** [comment]
No problem :smile: I'm just not sure how big the problem is and if its only relevant for Strimzi users because of the naming convention with the -> or if all MM2 users are affected.

**Comment by scholzj** [comment]
I think that depends. Every Strimzi MM2 user will use these names whcih cause the URL issue. But outside of Strimzi it might not necessarily be the case.

**Comment by AljoschaP** [comment]
I think we can close the issue or what do you think?

**Comment by scholzj** [comment]
Yeah, thanks for the investigation!

**Comment by mimaison** [comment]
Good catch! I also had completely forgotten about this one.

**Comment by douglasawh** [comment]
Is this fixed or is it just closed as a Kafka issue? I see that there was *a* fix for Kafka 2.8.1 and 3.0, but KAFKA-9981 is still open.

I see this comment: "yep, agreed - https://issues.apache.org/jira/browse/KAFKA-9981 appears to only affect running MM2 in distributed cluster mode, not via a Kafka Connect instance" but I am not sure I understand it.

Is the idea here that you could use a Kafka Connect cluster make MirrorMaker kinda work? This was opened as "multiple MM2 worker nodes" so I am just confused as to how exactly Kafka Connect is relevant.
